HIP/hip-792.md

```diff
@@ -0,0 +1,290 @@
+
+This will aid developers who were limited to `ECRECOVER` authorization flows, and `hederaAccountService.isAuthorized(address, messageHash, signatureBlob)` flows, who will now be able to expand authorization checks to include smart contract ID based authorization.
+
+| hash | signature | return | description |
+| --- | --- | --- | --- |
+|  | isAuthorizedCurrentTransaction() | bool | `true` if account is authorized to carry out transaction execution on account. Accepts protobuf key signature blobs. May be used for ECDSA, EdDSA simple key flows, and complex key flows which include any of ECDSA keys, EdDSA keys, and smart contract IDs. |
+
+### `isAuthorizedCurrentTransaction()` Function Usage
+
+This function behaves identically to `isAuthorized(address, messageHash, signatureBlob)` as defined in HIP-632, with the following key differences:
+
+- It is called without specifying any parameters
+- This function extracts the values that it needs in order to validate if a transaction is authorized from the current transaction
+- Therefore it designed to be used exclusively on the current transaction, which is still in-flight
+  - This is clearly communicated by the `CurrentTransaction` suffix in the function name
+
+No new protocol buffer schema definitions are needed as there are no parameters for this function.
+Internal protocol buffer schema definitions that need to be used to process this function
+would be the existing ones already present in Hedera's base account system,
+such as `Key`, `ContractID`,  `KeyList` and `ThresholdKey`. Potentially, this function may also use `SignatureMap` and `SignaturePair`,
+as defined in HIP-632, if deemed necessary during implementation.
+
+### Example usage flows
+
+Happy path example:
+
+- Account `0.0.12345` has a `ThresholdKey` of ([`customSc`, `edDsaKey`, `ecDsaKey`], 3) --> transactions need to be authorised by all 3.
+- `customSc` is a smart contract that has been deployed to HSCS
+- `customSc` contains a function `foo` which invokes `hederaAcountService.isAuthorizedCurrentTransaction()`
+- Transaction `tx` with `customSc.foo` invocation is instantiated, client-side
+- `tx` is signed by `edDsaKey`, client-side
+- `tx` is also subsequently signed by `ecDsaKey`, client-side
+- `tx` is executed by submitting it to the network
+- HSCS' EVM parses `tx` and invokes the function `customSc.foo`
+- This in turn invokes the system contract function
+  `hederaAcountService.isAuthorizedCurrentTransaction()`
+- `hederaAcountService` invokes the system contract implementation,
+  which uses `TxnAwareEvmSigsVerifier` to determine that `tx` is authorised, because:
+  - `tx` has been signed by `edDsaKey`, which is a member of account `0.0.12345`'s `ThresholdKey`
+  - `tx` has also been signed by `ecDsaKey`, which is a member of account `0.0.12345`'s `ThresholdKey`
+  - `tx` invokes `customSc`, which is a member of account `0.0.12345`'s `ThresholdKey`
```

Comment:
Given this, a contract would well have attempted the transaction and based on a success or error response code it would know.
The difference here would be contract flow and maybe cost in the query cost vs the failed transaction.
Given the current gas strategy the incentive for this seems to be more around cost efficiency

Reply:
If the cryptographic keys are not provided, and this key requires 3-of-3,
we know (before even getting to the smart contract invocation) that the best case scenario is to get to 1-of-3,
so this would ultimately be destined to unauthorised.
In such a case (and others of a similar vein) the implementation can perform a pre-check, and "fail fast".

> With the cryptographic key signatures are not provided in the parent transaction and if there are more than 1 contractIDs then this method will always return false.

I think this would apply for an account with a different key requirement, for example `ThresholdKey([edDsaKey, ecDsaKey, scId], 1)`.
In such a case, if the transaction is not signed by any of the cryptographic keys,
since the requirement is 1-of-3, and it is still possible to get to 1-of-3 by executing the smart contract,
then it would need to proceed with the execution attempt.

> Given this, a contract would well have attempted the transaction and based on a success or error response code it would know.

Yes, you're right - I haven't quite considered execution cost (and optimisation) yet.
I think we can use the cost of the verification that the HTS system contract presently has as a starting point.

> The difference here would be contract flow and maybe cost in the query cost vs the failed transaction.
Given the current gas strategy the incentive for this seems to be more around cost efficiency

Reply:
> I think this would apply for an account with a different key requirement, for example ThresholdKey([edDsaKey, ecDsaKey, scId], 1).
> In such a case, if the transaction is not signed by any of the cryptographic keys,
> since the requirement is 1-of-3, and it is still possible to get to 1-of-3 by executing the smart contract,
> then it would need to proceed with the execution attempt.

Correct what i mean is the check will only even return true if all cryptographic requirements are met whether 0, 1 or n.

Reply:
Other point is why query when if the only time you're get true is if you were actually authorized.
Only other consideration is price but that might be case by gas based on gasLimit and how  expensive the operation is vs the query.

Reply:
Querying ahead is nice if possible in many scenarios though.